src/clj/puppetlabs/http/client/async.clj

```diff
@@ -192,16 +191,18 @@
   [opts http-response]
   (let [headers       (get-resp-headers http-response)
         orig-encoding (headers "content-encoding")]
-    {:opts                  opts
+    {:opts opts
      :orig-content-encoding orig-encoding
-     :status                (.. http-response getStatusLine getStatusCode)
-     :headers               headers
-     :content-type          (parse-content-type (headers "content-type"))
-     :body                  (do
-		              (if (:decompress-body opts)
-		                (.process (ResponseContentEncoding.) http-response nil))
-		              (when-let [entity (.getEntity http-response)]
-		                (.getContent entity)))}))
+     :status (.. http-response getStatusLine getStatusCode)
+     :headers headers
+     :content-type (parse-content-type (headers "content-type"))
+     :body (do
+             (if (:decompress-body opts)
+               (.process (ResponseContentEncoding.)
+                         http-response
+                         (BasicHttpContext.)))
```

Comment:
Yeah, this doesn't seem right to me either.  It seems like we're calling `.process` for the side-effect of having decompression done properly for the response body but that this only works because the default `RequestConfig` that gets derived from the empty `BasicHttpContext` happens to have `contentCompressionEnabled` set to `true` by default.  I suppose that default isn't likely to change, but the multiple levels of implicitly derived defaults doesn't feel right.

I suppose this change isn't any worse than what we're getting in current code.  I found it interesting that the handling for the `JavaClient` is completely different than for the Clojure layer:

https://github.com/puppetlabs/clj-http-client/blob/http-client-0.4.2/src/java/com/puppetlabs/http/client/impl/JavaClient.java#L238-L240
https://github.com/puppetlabs/clj-http-client/blob/http-client-0.4.2/src/java/com/puppetlabs/http/client/impl/JavaClient.java#L360-L375

The Java client implementation seems a lot more intentional about how compression is handled for the response, although it may be redundant with logic that we can get more "for free" with some of the Apache helper methods?

Would be nice if we could end up with an implementation where we know that the same logic is applied for both the Clojure and Java client implementations.  Could lump some of this into the work for [TK-179](https://tickets.puppetlabs.com/browse/TK-179) if we ever get back around to that, but maybe a separate ticket tracking the extra work that we're talking about here would be good at a minimum.